cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,293 @@
+        size = size.expandedTo(item->sizeHint());
+    }
+    return size.isValid() ? size : QSize(0, 0);
+}
+
+/**
+ * @brief Returns the minimum size required to display all layout items.
+ * @return The minimum QSize needed by the layout.
+ */
+QSize FlowLayout::minimumSize() const
+{
+    QSize size;
+    for (QLayoutItem *item : items) {
+        size = size.expandedTo(item->minimumSize());
+    }
+
+    size.setWidth(qMin(size.width(), getParentScrollAreaWidth()));
+    size.setHeight(qMin(size.height(), getParentScrollAreaHeight()));
+
+    return size.isValid() ? size : QSize(0, 0);
+}
+
+/**
+ * @brief Adds a new item to the layout.
+ * @param item The layout item to add.
+ */
+void FlowLayout::addItem(QLayoutItem *item)
+{
+    items.append(item);
+}
+
+/**
+ * @brief Retrieves the count of items in the layout.
+ * @return The number of layout items.
+ */
+int FlowLayout::count() const
+{
+    return items.size();
+}
+
+/**
+ * @brief Returns the layout item at the specified index.
+ * @param index The index of the item to retrieve.
+ * @return A pointer to the item at the specified index, or nullptr if out of range.
+ */
+QLayoutItem *FlowLayout::itemAt(int index) const
+{
+    return items.value(index);
+}
+
+/**
+ * @brief Removes and returns the item at the specified index.
+ * @param index The index of the item to remove.
+ * @return A pointer to the removed item, or nullptr if out of range.
+ */
+QLayoutItem *FlowLayout::takeAt(int index)
+{
+    return (index >= 0 && index < items.size()) ? items.takeAt(index) : nullptr;
+}
+
+/**
+ * @brief Gets the horizontal spacing between items.
+ * @return The horizontal spacing if set, otherwise a smart default.
+ */
+int FlowLayout::horizontalSpacing() const
+{
+    return (m_hSpace >= 0) ? m_hSpace : smartSpacing(QStyle::PM_LayoutHorizontalSpacing);
+}
+
+/**
+ * @brief Gets the vertical spacing between items.
+ * @return The vertical spacing if set, otherwise a smart default.
+ */
+int FlowLayout::verticalSpacing() const
+{
+    return (m_vSpace >= 0) ? m_vSpace : smartSpacing(QStyle::PM_LayoutVerticalSpacing);
+}
+
+/**
+ * @brief Calculates smart spacing based on the parent widget style.
+ * @param pm The pixel metric to calculate.
+ * @return The calculated spacing value.
+ */
+int FlowLayout::smartSpacing(QStyle::PixelMetric pm) const
+{
+    QObject *parent = this->parent();
+    if (!parent) {
+        return -1;
+    } else if (parent->isWidgetType()) {
+        QWidget *pw = static_cast<QWidget *>(parent);
+        return pw->style()->pixelMetric(pm, nullptr, pw);
+    } else {
+        return static_cast<QLayout *>(parent)->spacing();
+    }
+}
+
+/**
+ * @brief Gets the width of the parent scroll area, if any.
+ * @return The width of the scroll area's viewport, or 0 if not found.
+ */
+int FlowLayout::getParentScrollAreaWidth() const
+{
+    QWidget *parent = parentWidget();
+    while (parent) {
+        if (QScrollArea *scrollArea = qobject_cast<QScrollArea *>(parent)) {
```

Comment:
nullptr check

Reply:
while (parent) should perform nullptr checking, since afaik, it'll fail as soon as it hits a nullptr.